docs/keps/20200619-federated-resource-status.md

```diff
@@ -0,0 +1,408 @@
+The value of `status.clusters.status` can be extracted from the `"ReplicaFailure"`
+status of the `Deployment` and so reused for visibility.
+
+Likewise, the status of the `FederatedDeployment` remains `Ready=True` in the rest
+of clusters: `cluster1` and `cluster2`.
+
+If a federated resource does not have a status field, a successful creation/update would
+reflect its readiness. Then the `ReadyCondition` would be satisfied by its creation.
+For these resources the value of `Conditions` would rely on the value of the propagation state of that resource.
+An example could be a `ClusterRole` resource that doesn't have a status property, but
+kubefed nowadays reports if the propagation of that resource worked.
+
+```yaml
+apiVersion: types.kubefed.io/v1beta1
+kind: FederatedClusterRole
+metadata:
+  name: test-clusterrole
+spec:
+  template:
+    rules:
+    - apiGroups:
+      - '*'
+      resources:
+      - '*'
+      verbs:
+      - '*'
+  placement:
+    clusters:
+    - name: cluster2
+    - name: cluster1
+status:
+  conditions:
+  - lastTransitionTime: "2020-05-25T19:47:59Z"
+    lastUpdateTime: "2020-05-25T19:47:59Z"
+    status: "True"
+    type: Propagation
+```
+
+However, there is a problem with this approach, the status schema varies based on the custom
+resource. Unfortunately that brings a problem when determining if a federated resource
+is ready or not.
+
+In the following, there is a list of Status objects of different custom resource definitions:
+
+```go
+type AddonStatus struct {
+	Ready bool          `json:"ready" yaml:"ready"`
+	Stage status.Status `json:"stage,omitempty" yaml:"stage,omitempty"`
+}
+
+type PodStatus struct {
+  Phase PodPhase
+ ...
+ }
+
+ type ServiceStatus struct {
+  LoadBalancer LoadBalancerStatus
+ }
+
+// KonvoyClusterStatus defines the observed state of KonvoyCluster
+type KonvoyClusterStatus struct {
+	// Phase represents the current phase of Konvoy cluster actuation.
+	// E.g. Pending, Provisioning, Provisioned, Deleting, Failed, etc.
+	// +optional
+	Phase KonvoyClusterPhase `json:"phase,omitempty"`
+
+  ...
+}
+```
+
+As mentioned, their Status schema is quite different from one to another.
+
+Consequently, the intention is to **enforce** in the federated resources this approved [recommendation](https://github.com/kubernetes/enhancements/pull/1624/files).
+to expect a common schema for `.status.conditions` and share golang logic for common Get, Set, Is for `.status.conditions`.
+
+1. For all new APIs, have a common type for `.status.conditions`.
+2. Provide common utility methods for `HasCondition`, `IsConditionTrue`, `SetCondition`, etc.
+3. Provide recommended defaulting functions that set required fields and can be embedded into conversion/default functions.
+
+By following this approach, kubefed would be able to properly consume and report the status
+of any federated resource by checking the `status.conditions` (e.g `Ready=True`) fields.
```

Comment:
Do you mean `type GenericCondition struct` should be replaced by the standard condition?
I think it's totally doable given that we are still in beta state.

Reply:
Yes, that is the intention, or at least, to extend it.